### ISSUE TITLE
useless_asref message not accurate for mut slices

### ISSUE THREAD

**Issue by ijackson** [rust-lang/rust-clippy#5862]
Hi.  (Thanks for clippy.  I just discovered it and I found it really quite helpful - it found several bugs and a number of simplifications in the small project I have just run it on.)

Consider this code:

```rust
use std::io;
use std::io::{Read,Write};

struct Thing;

impl Read for Thing {
    fn read(&mut self, orig_buf: &mut [u8]) -> Result<usize,io::Error> {
        let orig_wanted = orig_buf.len();
        #[allow(clippy::useless_asref)]
        let mut buf = orig_buf.as_mut();
        // without the .as_mut():
        /*
11 |     fn read(&mut self, orig_buf: &mut [u8]) -> Result<usize,io::Error> {
   |                        -------- move occurs because `orig_buf` has type `&mut [u8]`, which does not implement the `Copy` trait
...
14 |         let mut buf = orig_buf;
   |                       -------- value moved here
...
21 |         dbg!(&orig_buf[0..generated]);
   |               ^^^^^^^^ value borrowed here after move
        */
        
        // imagine more complex write sequence
        buf.write_all(&[ 1, 2 ])?;

        let generated = orig_wanted - buf.len();
        dbg!(&orig_buf[0..generated]);
        Ok(generated)
    }
}

```

I fed it to clippy.  Without the `#[allow]`, clippy issues a warning for `orig_buf.as_mut`,
```
warning: this call to `as_mut` does nothing
  --> src/lib.rs:13:23
   |
13 |         let mut buf = orig_buf.as_mut();
   |                       ^^^^^^^^^^^^^^^^^ help: try this: `orig_buf`
   |
```

But that's not right.  Effectively, `.as_mut()` is "cloning" the reference by reborrowing the underlying slice.

Possibly `.as_mut()` is poor style for this and this would be preferred style:
```
        let mut buf = &mut *orig_buf;
```

In which case suggesting this in those kind of cases would be better.  I'm not sure this problem is limited to slices.  It seems to exist (at least) for any mutable reference-like thing, where `as_ref` returns a newly mutable reference to the same underlying data.

### Meta

- `cargo clippy -V`: clippy 0.0.212 (6c8927b 2020-07-26)
- `rustc -Vv`:
```
rustc 1.47.0-nightly (6c8927b0c 2020-07-26)
binary: rustc
commit-hash: 6c8927b0cf80ceee19386026cf9d7fd4fd9d486f
commit-date: 2020-07-26
host: x86_64-unknown-linux-gnu
release: 1.47.0-nightly
LLVM version: 10.0
  ```

**Comment by divagant-martian** [comment]
I'd like to start working on this, any tips?

**Comment by divagant-martian** [comment]
@oli-obk saw you gave some comments on #3480. Would you know what's going on here? Tbh I can't differentiate between this case and one where the call to `as_mut` is truly useless

**Comment by oli-obk** [comment]
hmm... this reminds me a bit of our "useless clone" lint. Basically it is only useless, if the value being as_mut/as_ref'ed is ... uh... i don't know which word it is, but something-variant (co-, contra-, in-, ... I have to read up). Anyway, if you have a mutable reference or a struct containing a mutable reference, it's all the same, you need to reborrow, otherwise you'll be moving out of it. Now... if there are no more uses after the .as_mut, then it is useless, otherwise it is not. this is like with "useless clone", a clone is useless if there are no more uses of the value later.

So, maybe we should integrate this lint into the useless clone lint, or generalize the logic for useless clone so we can re-use it for other lints (the latter may be better).

**Comment by profetia** [comment]
I can no longer reproduce this in the latest nightly. Maybe this is already fixed.
